File: packages/provide-uterm-server/src/provide/uterm/tunnel/token_hash.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Hash-based at-rest storage for tunnel bearer tokens.

The hub holds tunnel tokens in process memory. A memory disclosure (core
dump, debugger attach, log of object state) on the server would otherwise
leak every active share/control token verbatim. Hashing the tokens before
storage means a disclosure leaks only digests; the original tokens cannot
be reconstructed without brute-forcing the BLAKE2b preimage, which is
infeasible for the 256-bit ``secrets.token_urlsafe(32)`` values that
``routes/tunnels.py`` issues.

Two-call API: ``hash_token`` is used at issuance time before storing the
hash; ``verify_token`` is used at every authentication site to compare a
caller-supplied plain token against the stored hash, in constant time.
"""

from __future__ import annotations

import hashlib
import secrets

# 32-byte digest is the BLAKE2b output that matches the entropy of the
# 32-byte urlsafe token we issue. Larger is wasteful, smaller is weaker.
_DIGEST_BYTES = 32


def hash_token(plain: str) -> str:
    """Return the BLAKE2b hex digest of ``plain``.

    Returns the empty string for an empty/None token so callers can treat
    "no token configured" the same as "no match".
    """
    if not plain:
        return ""
    return hashlib.blake2b(plain.encode("utf-8"), digest_size=_DIGEST_BYTES).hexdigest()


def verify_token(plain: str, stored_hash: str) -> bool:
    """Constant-time compare ``plain``'s hash against ``stored_hash``.

    Both the empty-stored-hash case and the empty-plain case return False —
    a configured-but-empty slot must never authenticate any caller.
    """
    if not plain or not stored_hash:
        return False
    candidate = hash_token(plain)
    return secrets.compare_digest(candidate, stored_hash)

File: packages/provide-uterm-ts/src/tunnel/token-hash.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Hash-based at-rest storage for tunnel bearer tokens.
 *
 * Port of `provide.uterm.tunnel.token_hash`. The hub holds tunnel tokens in
 * process memory, so a memory disclosure — a core dump, a debugger, a log of
 * object state — would otherwise leak every active share and control token
 * verbatim. Storing the digest means a disclosure leaks digests, which redeem
 * nothing: reconstructing the token means brute-forcing a BLAKE2b preimage of
 * a 256-bit value.
 *
 * Two calls: {@link hashToken} at issuance, {@link verifyToken} at every
 * authentication site.
 */

import { timingSafeEqual } from "node:crypto";
import { blake2b } from "@noble/hashes/blake2.js";
import { bytesToHex } from "@noble/hashes/utils.js";

/**
 * The digest length the reference uses, matching the entropy of the 32-byte
 * token it issues. Larger is wasteful, smaller is weaker.
 */
const DIGEST_BYTES = 32;

/**
 * The BLAKE2b hex digest of `plain`.
 *
 * Empty for an empty token, so a caller can treat "no token configured" the
 * same as "no match".
 *
 * BLAKE2b-256 rather than Node's `createHash`, which offers only
 * BLAKE2b-512 — and truncating that would give a different digest, since the
 * output length is mixed into BLAKE2b's parameter block.
 */
export function hashToken(plain: string): string {
  if (plain === "") {
    return "";
  }
  return bytesToHex(blake2b(new TextEncoder().encode(plain), { dkLen: DIGEST_BYTES }));
}

/**
 * Whether `plain` hashes to `storedHash`, compared in constant time.
 *
 * Both an empty token and an empty stored hash are refused: a
 * configured-but-empty slot must never authenticate any caller.
 */
export function verifyToken(plain: string, storedHash: string): boolean {
  // Either half of this guard is redundant while the other stands — an empty
  // token hashes to nothing and an empty slot holds nothing, so the length
  // check below refuses both anyway. Stated in full because dropping *both*
  // is the hole: two empty strings would then compare equal.
  if (plain === "" || storedHash === "") {
    return false;
  }
  const candidate = Buffer.from(hashToken(plain), "utf8");
  const stored = Buffer.from(storedHash, "utf8");
  // `timingSafeEqual` raises on a length mismatch rather than answering, and
  // a store can hold something the wrong length. The length is not the secret.
  if (candidate.length !== stored.length) {
    return false;
  }
  // The reason this is not `candidate === storedHash`: the two answer alike,
  // and no test can tell them apart. What differs is how long the wrong
  // answer takes.
  return timingSafeEqual(candidate, stored);
}
